### ISSUE TITLE
Improved splitting method

### ISSUE THREAD

**Issue by peastman** [deepchem/deepchem#2324]
I'm opening this as a place to describe something I'm working on and get feedback about it.

We recently changed the molnet loader functions to use scaffold splitting instead of random splitting, which is a big improvement.  But there's room for further improvement.  On small datasets I've found that Butina splitting works even better.  Unfortunately, it crashes on larger datasets.  I think that's because its memory use scales as O(n^2) in the number of molecules.

ButinaSplitter works by clustering the molecules based on their fingerprints and keeping each cluster together.  So my first idea was to try other clustering methods and look for one that scales better.  Scikit-learn and SciPy both have a variety of clustering algorithms.  Unfortunately it appears that almost all of them either have memory that scales as O(n^2), or else are hardcoded to use Euclidean distance which [doesn't work very well](https://jcheminf.biomedcentral.com/articles/10.1186/s13321-015-0069-3) for fingerprints.  So far I've been unable to find one that meets all of our needs.

Thinking about it more, I'm not sure that clustering is really the right way to approach this.  It might be better to treat it as something more like a weighted maxcut problem: the goal is to divide the samples into two groups while maximizing the distance between pairs that are in different groups.  Viewed that way, I think I can come up with a method that produces good results, only has linear memory use, and is fast enough to be useful even for large datasets.

**Comment by peastman** [comment]
I came up with a very simple algorithm that seems to work quite well.  It adds molecules one at a time, alternating between the two groups.  At each step it does the following:

1. Identify the unassigned molecule that is least similar to anything in one group.
2. Add it to the other group.

To test it, I compared several splitting methods: random, scaffold, butina, and fingerprint (my new method).  For each one I trained a MultitaskClassifier on the training set then computed the ROC AUC on the training and validation sets.  The training set score should be essentially the same for all of them.  The better the splitter is working, the lower the validation set score will be.  I also recorded the total time to load the dataset (in seconds).

First for BACE which has only 1513 molecules:

|Splitter|Training|Validation|Time|
|---|---|---|---|
|random|0.991|0.878|4|
|scaffold|0.988|0.722|5|
|butina|0.989|0.714|5|
|fingerprint|0.988|0.577|5|

Next for Tox21 which is a bit bigger with 7831 molecules:

|Splitter|Training|Validation|Time|
|---|---|---|---|
|random|0.969|0.788|16|
|scaffold|0.971|0.702|19|
|butina|0.970|0.670|28|
|fingerprint|0.968|0.611|37|

And finally HIV which has 41,127 molecules.  This one is too big to use Butina.  It hangs for about half an hour and then my computer crashes.

|Splitter|Training|Validation|Time|
|---|---|---|---|
|random|0.996|0.789|91|
|scaffold|0.995|0.764|117|
|fingerprint|0.997|0.574|797|

In all cases it's a big improvement over scaffold splitting.  In fact, the validation set performance is only slightly better than totally random.  I wonder if it's splitting *too* effectively, by ensuring the validation set doesn't contain anything that at all resembles anything in the training set?

I'd like to see if I can speed it up a bit more, but as long as you're patient it can already handle moderately large datasets in an amount of time that isn't too absurd.

**Comment by ncfrey** [comment]
This is really cool! It does seem like your fingerprint method is creating a situation where the validation set is out of distribution. For materials datasets it's uncommon to do any kind of "scaffold" splitting. Some papers have recommended [k-means clustering](https://pubs.rsc.org/ko/content/articlehtml/2018/me/c8me00012c) for small datasets and I've used that, but that runs into the O(n^2) problem. I think fingerprint splitting would be really valuable for materials as well, and would relate to a problem I've been interested in, trying to construct standards for material fingerprints.

I guess in my mind these are two different cases: you might only be interested in the interpolative power of your model, in which case the fingerprint splitter is not actually constructing a good validation set. If you're interested in extrapolative out-of-distribution detection, then fingerprint splitter seems better than any of these other methods! If the intent of scaffold splitting is to be somewhere in-between and also computationally efficient, maybe that makes sense as the default. Then fingerprint splitting could be recommended for more robust OOD validation.

**Comment by rbharath** [comment]
This is super cool! This is a robust separator that seems to really stress test out-of-distribution behavior. I've noticed in live deployments, scaffold split often still underestimates true experimental performance, so I think there would be a lot of value in this new fingerprint splitter for applications. 

As a a couple of thoughts/comments:

1. It might be useful to see if we can run the new fingerprint algorithm on datasets with up to 500K molecules. I think increasingly, datasets of this size are starting to come online (within companies, but not quite yet open source).
2. Echoing @ncfrey, this would be really cool to demonstrate for materials! I think the methodology is broadly applicable. We could do this for protein-ligand systems using the new interaction fingerprints as well.
3. I think the combination of all these ideas could make for a really nice paper introducing the new algorithm and testing on large molecular datasets, materials datasets, and protein-ligand datasets.

**Comment by peastman** [comment]
This raises the question of how to evaluate splitting methods.  Looking at some papers, they mostly seem to assume that a "harder" splitting is always better.  I don't think that's necessarily true.  If you make it too hard, that won't give a good representation of the real quality of your model.  As an extreme example, if a particular bit of the fingerprint is never set for any training sample, there's no way the model can possibly learn how to interpret it.  That doesn't reflect on the quality of the model.  It just means you're asking it to solve an impossible learning task.

That made me think we should look at the individual bits within the fingerprint.  Are we asking the model to predict based on features it's never seen before, or just novel combinations of familiar features?

As a simple test, I computed the average of all fingerprints in each dataset.  That gives a vector telling us what fraction of all samples have each bit set.  Then I computed a few statistics:

1. The mean absolute difference between these vectors for the training and validation sets.
2. The number of bits (out of 1024) that are set in less than 1% of all training samples.

Here are the results.  First for BACE, in which fingerprint splitting was *much* harder than any of the others:

|Splitter|MAE|Bits|
|---|---|---|
|random|0.011|325|
|scaffold|0.020|343|
|butina|0.038|387|
|fingerprint|0.056|365|

For Tox21, which had a somewhat smaller gap between methods:

|Splitter|MAE|Bits|
|---|---|---|
|random|0.004|290|
|scaffold|0.018|406|
|butina|0.007|337|
|fingerprint|0.017|422|

And finally for HIV, which had minimal difference between random and scaffold splitting, but a huge difference between those and fingerprint splitting:

|Splitter|MAE|Bits|
|---|---|---|
|random|0.002|78|
|scaffold|0.007|102|
|fingerprint|0.019|104|

It appears that the number of bits that are set in less than 1% of training samples has little predictive power.  On the other hand, the MAE between the training and validation distributions is much more predictive.  This doesn't clearly say that fingerprint splitting is either a "fair" or "unfair" way of making the task harder.  But it does appear that at least it doesn't do it by starving the model of training data on individual features.

Another way of evaluating splitting methods is to take temporal splitting as being ideal, since it most closely reflects the way the model will be used in production.  We can then compare how closely other splitting methods reproduce its results.  [This paper](https://pubs.acs.org/doi/pdf/10.1021/acs.jcim.9b00237) took that approach.  Their results for different datasets are a bit mixed.  See figures 11, 12, and 13.  In most cases, temporal splitting is much harder than random.  Sometimes scaffold splitting is close to random, and sometimes it's closer to temporal.

**Comment by peastman** [comment]
I managed to speed it up a fair amount.  Loading HIV is down to 267 seconds.  500,000 molecules would still take several hours, but I think it ought to work.

I'll clean up the code and create a PR.

We already have a class called FingerprintSplitter, but it's way too slow to be useful (several minutes to split BACE), and its memory scales as O(n^2).  The way it works is actually closely related to this one, except it tries to maximize the similarity within each dataset instead of minimizing the similarity between datasets.  Is it ok if I get rid of that class and make this the new FingerprintSplitter class?

**Comment by peastman** [comment]
Closing this since the new splitting has been merged.
